post-repo-inmem/src/main/java/klon/post/repo/inmem/PostRepositoryInmemImpl.java

```diff
@@ -0,0 +1,52 @@
+package klon.post.repo.inmem;
+
+import klon.post.repo.api.Post;
+import klon.post.repo.api.PostRepository;
+import lombok.Builder;
+
+import java.util.*;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.stream.Stream;
+
+@Builder
+public class PostRepositoryInmemImpl implements PostRepository {
+
+    private final ConcurrentHashMap<String, List<InmemPost>> posts = new ConcurrentHashMap<>();
+
+    private final PostMapper mapper;
+
+    @Override
+    public void addPost(Post post) {
+        InmemPost inmemPost = mapper.toInmemPost(post);
+        posts.merge(post.getUserId(), Collections.synchronizedList(new ArrayList<>(Collections.singletonList(inmemPost))),
+                (inmemPosts, inmemPosts2) -> {
+                    inmemPosts.addAll(inmemPosts2);
+                    return inmemPosts;
+                });
+    }
+
+    @Override
+    public Optional<Post> getPost(String postId) {
+        return posts.values()
+                .stream()
+                .flatMap(Collection::stream)
+                .filter(post -> postId.equals(post.getPostId()))
+                .map(mapper::toPost)
+                .findFirst();
+    }
+
+    @Override
+    public Stream<Post> getPostsByUser(String userId) {
+        List<InmemPost> inmemPosts = posts.get(userId);
+        if (inmemPosts == null) {
+            return Stream.empty();
+        }
+        inmemPosts.sort(Comparator.comparing(InmemPost::getCreatedTime));
```

Comment:
Odnośnie tej synchronized list. Tu byłoby miejsce do synchronizacji, by sortowanie i dodawanie sobie nie wchodziły w paradę (nie wiem jak się zachowa sort przy jednoczesnym dodaniu do listy), ale może lepiej tutaj zrobić kopię i posortować kopię (nie wiem czy kopiowanie jest bezpieczne, jeśli jednocześnie dodajemy).
Być może Lista, która wychodzi z Collections.synchronizedList to zabezpiecza. **Nie wiem, więc tylko zgłaszam swoje wątpliwości**.
Uwaga wydajnościowa - lepiej sortować kopię, bo oryginał jest potencjalnie współdzielony z innymi wątkami i w ten czy inny sposób musi następować synchronizacja/blokowanie na czas sortowania.
A może w ogóle sortowanie wrzucić na koniec dodawania postu - będzie sumarycznie mniej operacji sortowania.

Reply:
> A może w ogóle sortowanie wrzucić na koniec dodawania postu - będzie sumarycznie mniej operacji sortowania.

Faktycznie, słuszna uwaga.

Reply:
> Być może Lista, która wychodzi z Collections.synchronizedList to zabezpiecza.

Z tego co piszą [tu](https://stackoverflow.com/a/4584654/2710264) to jednak nie zabezpiecza i kopię trzeba robić w synchronizowanym bloku.